src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,279 @@
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Format;
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.request.Regs;
+import com.iab.openrtb.response.Bid;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.criteo.model.CriteoGdprConsent;
+import org.prebid.server.bidder.criteo.model.CriteoPublisher;
+import org.prebid.server.bidder.criteo.model.CriteoRequest;
+import org.prebid.server.bidder.criteo.model.CriteoRequestSlot;
+import org.prebid.server.bidder.criteo.model.CriteoResponse;
+import org.prebid.server.bidder.criteo.model.CriteoResponseSlot;
+import org.prebid.server.bidder.criteo.model.CriteoUser;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.EncodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.ExtImpCriteo;
+import org.prebid.server.proto.openrtb.ext.request.ExtRegs;
+import org.prebid.server.proto.openrtb.ext.request.ExtUser;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpMethod;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CriteoBidder implements Bidder<CriteoRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpCriteo>> CRITEO_EXT_TYPE_REFERENCE =
+            new TypeReference<ExtPrebid<?, ExtImpCriteo>>() {
+            };
+
+    private final String endpointUrl;
+    private final JacksonMapper jsonMapper;
+    private final CriteoSlotIdGenerator criteoSlotIdGenerator;
+
+    public CriteoBidder(String endpointUrl, JacksonMapper jsonMapper, CriteoSlotIdGenerator criteoSlotIdGenerator) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.jsonMapper = Objects.requireNonNull(jsonMapper);
+        this.criteoSlotIdGenerator = criteoSlotIdGenerator;
+    }
+
+    @Override
+    public Result<List<HttpRequest<CriteoRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final CriteoRequest.CriteoRequestBuilder criteoRequestBuilder
+                = CriteoRequest.builder();
+
+        criteoRequestBuilder.id(bidRequest.getId());
+
+        Integer networkId = null;
+
+        if (CollectionUtils.isNotEmpty(bidRequest.getImp())) {
```

Comment:
I don't really like how code looks from this line to https://github.com/prebid/prebid-server-java/pull/1252/files#diff-1829dea7a3d64204556a8ca37cbbc302adcc12e8dd56cbe792278f328bfd007dR115
I just didn't found a way to extract to separate method yet

Reply:
btw, see `org.prebid.server.validation.RequestValidator#validate`:
```
if (CollectionUtils.isEmpty(bidRequest.getImp())) {
            throw new ValidationException("request.imp must contain at least one element");
}
```